application/classes/Ushahidi/Formatter/Post/CSV.php

```diff
@@ -37,113 +37,203 @@ public function __invoke($records)
-					}
-				}
-
-				// Assign post values
-				else
-				{
-					unset($record[$key]);
-					$this->assignRowValue($record, $key, $val);
-				}
+			$values = [];
+			foreach ($heading as $key => $value) {
+				$values[] = $this->getValueFromRecord($record, $key);
 			}
-
-			// Pad record
-			$missing_keys = array_diff(array_keys($heading), array_keys($record));
-			$record = array_merge($record, array_fill_keys($missing_keys, null));
-
-			// Sort the keys so that they match with columns from the CSV heading
-			ksort($record);
-
-			fputcsv($fp, $record);
+			fputcsv($fp, $values);
 		}
-
 		fclose($fp);
 
 		// No need for further processing
 		exit;
 	}
 
-	private function assignRowValue(&$record, $key, $value)
-	{
-		if (is_array($value))
-		{
-			// Assign in multiple columns
-			foreach ($value as $sub_key => $sub_value)
-			{
-				$record[$key.'.'.$sub_key] = $sub_value;
-			}
+	private function getValueFromRecord($record, $keyParam){
+		$return = '';
+		$keySet = explode('.', $keyParam); //contains key + index of the key, if any
+		$headingKey = $keySet[0];
+		$key = isset($keySet[1]) ? $keySet[1] : null;
+		$recordValue = isset ($record['attributes']) && isset($record['attributes'][$headingKey])? $record['values']: $record;
+		if($key === 'lat' || $key === 'lon'){
+			/*
+			 * Lat/Lon are never multivalue fields so we can get the first index  only
+			 */
+			$return = isset($recordValue[$headingKey][0][$key])? ($recordValue[$headingKey][0][$key]): '';
+		} else if ($key !== null) {
+			/**
+			 * we work with multiple posts which means our actual count($record[$key])
+			 * value might not exist in all of the posts we are posting in the CSV
+			 */
+			$return = isset($recordValue[$headingKey][$key])? ($recordValue[$headingKey][$key]): '';
+		} else{
+			$emptyRecord = !isset($record[$headingKey]) || (is_array($record[$headingKey]) && empty($record[$headingKey]));
+			$return = $emptyRecord ? '' : $record[$headingKey];
 		}
+		return $return;
+	}
 
-		// ... else assign value as single string
-		else
-		{
-			$record[$key] = $value;
+	/**
+	 * @param $fields: an array with the form: ["key": (value)] where value can be anything that the user chose.
+	 * @return array of sorted fields with a zero based index. Multivalue keys have the format keyxyz.index index being an arbitrary count of the amount of fields.
+	 */
+	private function createSortedHeading($fields){
+		$headingResult = [];
+		// fieldsWithPriorityValue: an associative array with the form ["uuid"=>[label: string, priority: number, stage: number],"uuid"=>[label: string, priority: number, stage: number]]
+		$fieldsWithPriorityValue = [];
+		/**
+		 * Separate by fields that have custom priority and fields that do not have custom priority assigned
+		 */
+		foreach ($fields as $fieldKey => $fieldAttr) {
+			if (!is_array($fieldAttr)) {
+				$headingResult[$fieldKey] = $fieldAttr;
+			} else if (is_array($fieldAttr) && isset($fieldAttr['nativeField'])){
```

Comment:
I think you can drop the is_array check because isset will test anything for the presence of a key and the previous if has already confirmed that this must be an array to reach this line.